Add rulesets to phpcs.xml, correct some phpcs issues from them

kienstra opened the pull request:
* Adds more rulesets to `phpcs.xml`, including checks for PHP compatibility and `'block-lab'` textdomain arguments for functions like `__()`
* Adds PHP 7.3 and 7.4 to the Travis build matrix
* Converts `array()` syntax to `[]`

kienstra committed:
Add rulesets to phpcs.xml, correct some phpcs issues from them

This tests PHP compatibility,
and ensures that __() has a second
(textdomain) argument.

kienstra commented on composer.json:
```diff
@@ -9,8 +9,10 @@
     },
     "require-dev": {
         "brain/monkey": "^2",
+        "dealerdirect/phpcodesniffer-composer-installer": "0.5.0",
```
Without this, sometimes there are error messages when doing `npm run lint:php`, like:

>ERROR: Referenced sniff "WordPress" does not exist

kienstra commented on phpcs.xml:
```diff
@@ -1,22 +1,39 @@
+	<exclude-pattern>*/vendor/*</exclude-pattern>
+	<exclude-pattern>*/package/*</exclude-pattern>
+	<exclude-pattern>*.js</exclude-pattern>
+	<exclude-pattern>*.css</exclude-pattern>
```
A lot of this diff is converting the file from spaces to tabs.

kienstra committed:
Use [] array syntax instead of array()

"As of PHP 5.4 you can also use the short array syntax, which replaces array() with []."

@see https://www.php.net/manual/en/language.types.array.php

kienstra commented on php/admin/class-settings.php:
```diff
@@ -122,7 +122,7 @@ public function prepare_notice( $notice ) {
 	 */
 	public function show_notices() {
-		$notices = get_option( 'block_lab_notices', array() );
+		$notices = get_option( 'block_lab_notices', [] );
```
I realized that by changing `phpcs.xml`, doing `npm run lint:php:fix` can convert all of the `array()` syntax to `[]`.

This produced a huge diff, and will of course need regression testing.

But I think it's important to use this more modern syntax. The `[]` syntax [started in PHP 5.4](https://www.php.net/manual/en/language.types.array.php), and we only support PHP 5.6+.

kienstra committed:
Change the ruleset name to be specific to Block Lab

Before, this only mentioned 'Plugins.'
Also, remove an extra indentation.

kienstra committed:
Remove extra indendation in array

This doesn't need to be indendented,
it can be on a single line.

kienstra committed:
Remove another extra array indentation

It's not needed to indent this,
it can be on a single line.

kienstra commented:
Hi @lukecarbis, great talking to you earlier.

Do you like the `[]` syntax here instead of `array()`? 

I think it's important to use the `[]` syntax for the plugin to be more modern, but that may be a personal preference.

kienstra committed:
Add PHP 7.3 and 7.4 to the .travis.yml build matrix

It's important to test these,
especially 7.3.

kienstra committed:
Attempt to address failed PHP 7.4 build

Using something from the
AMP plugin's .travis.yml.

kienstra committed:
Attempt to address failed Travis build

Update the PHPUnit version
in response to the error:
PHP Deprecated:  The each() function is deprecated. This message will be suppressed on further calls in /home/travis/.composer/vendor/phpunit/phpunit/src/Util/Getopt.php on line 38

kienstra committed:
Add a of directory to the require statement

This didn't have a directory before,
so it shouldn't work.

kienstra committed:
Change the PHPUNIT_VERSION to 7.5.17

In response to a failed build:
ERROR: Looks like you're using PHPUnit 8.4.3. WordPress is currently only compatible with PHPUnit up to 7.x.
Please use the latest PHPUnit version from the 7.x branch.

kienstra committed:
Attempt to prevent error in test from array access

2) Test_Template_Output::test_block_template
Trying to access array offset on value of type bool

lukecarbis commented:
I prefer `array()`, but only because that's what I've always used / seen / been taught.

I'm all for making the plugin more modern, though. I have seen `[]` around the place. Could you point me to a couple of good examples of modern WordPress plugins that prefer this syntax?

lukecarbis commented:
Also, WP core uses `array()`.

kienstra commented:
Right, Core does use `array()`. 

It looked strange to me also before the AMP plugin started using it:

https://github.com/ampproject/amp-wp/blob/09f1461b896c921d4afd2ffbeee1597043c890e8/includes/class-amp-theme-support.php#L105

I'm looking for more plugin examples.

lukecarbis commented:
Sorry, you'll have to find something more than a Google project to impress me. ;)

kienstra commented:
Haha 😄 

Yoast: https://github.com/Yoast/wordpress-seo/blob/9c43132323adfcc7e748a1f30af994c50b62f89e/inc/health-check.php#L149

kienstra commented:
Calypso uses `[]`: https://github.com/Automattic/wp-calypso/blob/fba2d80dc733828ff3fff47af86f0944cca83bff/apps/full-site-editing/full-site-editing-plugin/global-styles/includes/class-data-set.php#L20

To be honest, there are probably more plugins using `array()` than `[]`. [Jetpack](https://github.com/Automattic/jetpack/) (at least mainly) uses `array()`.

But I think it makes our plugin look more modern, and there's no concern for PHP supporting it, as we support 5.6+, while `[]` goes back to `5.4`. 

Still, I'm not passionate about this. It'd be easy to revert the one commit on this PR and discuss this separately.

lukecarbis commented:
Let's leave it in for now. I've asked around on Twitter and Slack, see what a few other people say. I'm also happy to be a trend-setter and just push for a new standard here too.

kienstra commented:
OK, sounds good. I need to do some good regression testing for this PR. It might be a few days, at least.

lukecarbis commented:
<img width="547" alt="Screen Shot 2019-12-05 at 8 27 59 am" src="https://user-images.githubusercontent.com/1097667/70187174-4ebff880-1739-11ea-83c5-089142b055d9.png">
<img width="667" alt="Screen Shot 2019-12-05 at 8 28 19 am" src="https://user-images.githubusercontent.com/1097667/70187178-4ff12580-1739-11ea-8733-3cb1399cb749.png">

Looks like the answer is pretty clear.

Worth mentioning this response, from @peterwilsoncc, on why this hasn't been adopted in WordPress core:

> It was reviewed but long form was decided upon for the reason that people with impaired vision can have trouble seeing the difference between [(

Also, @visser voted twice. 🙄

lukecarbis commented:
@kienstra If you put the letters WIP in the title of the PR (I usually use `[WIP]` at the start), we have a WIP check that will prevent merging this PR accidentally.

kienstra commented:
Nice, thanks for tweeting that poll!

kienstra commented:
>@kienstra If you put the letters WIP in the title of the PR (I usually use [WIP] at the start), we have a WIP check that will prevent merging this PR accidentally.

Good point that it's safer, as it blocks merging.

I've avoided doing that because there's no green check build icon, so at first glance it seems like the builds are failing. Of course, it's not hard to see that it's just because of the 'WIP' Travis check.

kienstra committed:
Remove duplication of compser global require statement

There's no need to repeat that.
So instead set PHPUNIT_VERSION to a default
if it isn't set.

kienstra committed:
Address Travis error, Remove the 'default' textdomain

This doesn't seem to be used
in Block Lab, and it can always be added
if it is ever used.

kienstra committed:
Again attempt to address the failed Travis build

I think I had the conditional wrong,
so that it was reassigning the PHPUNIT_VERSION
only if it already existed.

kienstra commented on php/views/license.php:
```diff
@@ -33,9 +33,9 @@
 					}
 
-					$expiry = date( get_option( 'date_format' ) );
+					$expiry = gmdate( get_option( 'date_format' ) );
```
This is one of the few changes to a PHP file in this PR, other than `array()` to `[]`.

This appears in 8 other places, and is in response to a PHPCS error.

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Could you please review this when you can?

* Adds more rulesets to `phpcs.xml`, including checks for PHP compatibility. 
* Adds PHP 7.3 and 7.4 to the `.travis.yml` build matrix
* Changes `array()` to `[]` 

Regression testing looks good. I created a new block, and it displayed and saved as expected in the block editor:

<img width="822" alt="new-block" src="https://user-images.githubusercontent.com/4063887/70590230-af1ac400-1b97-11ea-91ed-66c1b54b731b.png">

And a block with almost all of the fields saved its values and displayed them as expected:
<img width="824" alt="fields-block" src="https://user-images.githubusercontent.com/4063887/70590282-d2de0a00-1b97-11ea-8e2d-7baf4699713b.png">

Also, saving a license key worked as expected, and the Pro fields were available.

Thanks, Luke!

lukecarbis reviewed (approved):
It's going to take some time for me to get used to square bracket arrays, but that's okay! Looking good.

kienstra commented:
Thanks, @lukecarbis!
